### ISSUE TITLE
Document for how to configure pod communication cross  Kubernetes Clusters

### ISSUE THREAD

**Issue by gyliu513** [projectcalico/calico#1942]
Seems many people interested in this feature, we should have a document in Calico for how to setup.

This are [some discussion](https://github.com/projectcalico/calico/issues/908) and I still have some trouble to set up such clusters, but I can work on the document once cross cluster works.

/cc @caseydavenport @tmjd @karstensi

**Comment by caseydavenport** [comment]
@gyliu513 I think this would be a great thing to document and I'd be happy to help out with any gotchas. I'll read through the other thread.

The main thing is that there are a LOT of different ways this could be set up, so a one-size-fits-all solution probably doesn't exist. I think we'll need a doc which covers the main things you need to consider when setting up, and then with some examples would be great.

**Comment by gyliu513** [comment]
Thanks @caseydavenport , I want to start with calico router reflector, seems router reflector 0.5.0 will not work based on discussion here https://github.com/projectcalico/calico/issues/908#issuecomment-385160303 

I'm now trying with calico 2.6.6 and RR 0.4.2, will append more detail later for any issues.

**Comment by gyliu513** [comment]
Found an issue for 0.4.2 set up and opened an issue for my question at https://github.com/projectcalico/calico/issues/1948

**Comment by Camsteack** [comment]
Hey guys, any update on that ? We are looking at enabling pod-to-pod communication across Kubernetes cluster and it's difficult to find some documentation on that.

Cheers

**Comment by gyliu513** [comment]
@Camsteack  Please take a look at this document https://medium.com/ibm-cloud/multi-cluster-support-for-service-mesh-with-ibm-cloud-private-d7d791f9b778 for how to configure pod-to-pod communication across Kubernetes cluster with Node to Node Mesh.

**Comment by kesor** [comment]
@redbaron I was able to follow your guide and get the routes published in both clusters, but then the pods cannot actually reach each other because by default (in calico 3.7) IP-in-IP traffic from nodes that are not part of *this* cluster are automatically filtered using iptables.

There is an `ipset` called `cali40all-hosts-net` that includes the `FELIX_EXTERNALNODESCIDRLIST` and the server IPs of the current cluster, but it does not include the node IPs of the other cluster - thus traffic is dropped.

**Comment by ffilippopoulos** [comment]
@kesor you need to define the remote cluster ip pool like:
```
apiVersion: crd.projectcalico.org/v1
kind: IPPool
metadata:
  name: dev-gcp-pods
spec:
  cidr: 10.4.0.0/16
  ipipMode: CrossSubnet
  disabled: true
```
in each cluster and also you need to let felix know about the remote hosts subnet in `FELIX_EXTERNALNODESCIDRLIST`:
```
    env:
           - name: FELIX_EXTERNALNODESCIDRLIST
              value: '10.10.0.0/24'
```
in order for calico/bird to accept the routes and insert the correct paths into your route tables.
After that you might need network policies to allow the whole remote cluster pod ip range.
We are doing a similar approach to @redbaron solution but we are using `gobgp` instances because we need the flexibility of dynamic ip addresses for our cluster nodes.
We created a small blog post to describe it here: https://uw-labs.github.io/blog/kubernetes/2019/05/01/cross-cluster-comms.html

**Comment by kesor** [comment]
@ffilippopoulos following the guide by @redbaron worked well for us, the only missing piece was `FELIX_EXTERNALNODESCIDRLIST` which we wrongly assumed was a list of POD network CIDRs and not NODE networks :/

**Comment by tushar00jain** [comment]
@redbaron 's guide also worked for me, will also be good to have documentation on how to setup using route reflectors! Is it also possible to propagate service IP addresses so those could be routed as well?

**Comment by rgarcia89** [comment]
@redbaron is this still working for you with k8s v1.15.0?
In my test using the exact same manifest that I used for v.1.14.4 only the nodes can that are listed as primary path in the bgp routing table are able to reach the remote cluster...
Seems like something has changed.

**Comment by redbaron** [comment]
I haven updated to 1.15 yet, thanks for the heads up. If you find what changed please post it here

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by rgarcia89** [comment]
I just run another test. It seems like with 1.15 it only works when applying the "next hop keep" tweak that you mentioned above. Otherwise nodes that are not the next hop are not able to reach the remote cluster. This however was working without any issues in 1.14.4 also without the "next hop keep" tweak. Have you been able to reproduce it? 
I can send you the used manifests if you want. They are basically following your notes from above without the traffic optimization part

**Comment by rgarcia89** [comment]
@redbaron how do you advertise service ip addresses when using the " sed -ie "/multihop/a next hop keep;" /etc/calico/confd/templates/bird.cfg.template && exec /sbin/start_runit" tweak?

**Comment by redbaron** [comment]
I don't understand your question. Service with external traffic policy local works just fine with this setup.

-------- Original Message --------

**Comment by rgarcia89** [comment]
Yeah it was my fault - I totally forgot that I had to add the service address CIDR to the calico ippool...

**Comment by dimm0** [comment]
> I just run another test. It seems like with 1.15 it only works when applying the "next hop keep" tweak that you mentioned above. Otherwise nodes that are not the next hop are not able to reach the remote cluster. This however was working without any issues in 1.14.4 also without the "next hop keep" tweak. Have you been able to reproduce it?
> I can send you the used manifests if you want. They are basically following your notes from above without the traffic optimization part

Is there any better way to do this in 3.17?

**Comment by rgarcia89** [comment]
@dimm0 I have not seen any other way then to use this hack.
It is currently working on version v3.15 clusters.

Maybe @caseydavenport or @redbaron know more

**Comment by patjlm** [comment]
As far as I understood, BGP and IPIP encapsulation cannot be used on Microsoft Azure since their vnet is entirely controlling packet routing. We used vxlan for intra-cluster setup. Is calico cluster mesh working on azure ? If there are any doc on this, that'd be perfect !

**Comment by laxmanvallandas** [comment]
@redbaron , Is this guide still valid for v3.19.1. we tried on k8s 1.19. @tushar00jain ,  were there any tweaks in your setup besides above guide(incase, if you are still running the setup?)

**Comment by george-angel** [comment]
Hello @laxmanvallandas 

We are running a cross-cluster setup that heavily leverages calico, we have a write up here: https://uw-labs.github.io/blog/kubernetes,/multicluster/2021/07/21/kube-semaphore-intro.html

Hopefully that will be useful to you.

**Comment by laxmanvallandas** [comment]
@george-angel , That's exactly where i started with ;) . But, found that pod-to-pod communication is the prerequisite as mentioned [here](https://github.com/utilitywarehouse/semaphore-service-mirror#semaphore-service-mirror).

**Comment by ffilippopoulos** [comment]
@laxmanvallandas In this: https://uw-labs.github.io/blog/kubernetes,/multicluster/2021/07/21/kube-semaphore-intro.html pod to pod connectivity is achieved via a WireGuard mesh network created with: https://github.com/utilitywarehouse/semaphore-wireguard. This will configure local wg peers on all nodes that can route traffic for the respective remote pod sunbnets (assuming that you have node to node connectivity between cluster1 and cluster2). It is an alternative to the BGP route reflector approach described in this issue, with additional encryption provided from WireGuard. The service-mirroring and policies can be used regardless of the mean you select to flatten your pods subnets.

**Comment by laxmanvallandas** [comment]
@george-angel @ffilippopoulos Have raised an issue(may be not) that I am facing with semaphore-wireguard. 
https://github.com/utilitywarehouse/semaphore-wireguard/issues/22 . Can I get some inputs there?

**Comment by redbaron** [comment]
RIP

**Comment by caseydavenport** [comment]
Happy to keep things open, just comment!
